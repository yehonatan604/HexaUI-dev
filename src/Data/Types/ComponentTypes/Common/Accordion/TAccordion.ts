import { TComponent, ComponentOptions } from "../../TComponent";

export type TAccordion = TComponent<"div"> & {
    options: AccordionOptions;
}

export type AccordionOptions = ComponentOptions & {
    title: string;
    titlePadding?: string;
    titleVariant?: string;
    contentPadding?: string;
    chevronVariant?: string;
    chevronSize?: string;
    chevronDirection?: string;
    icon?: JSX.Element;
}