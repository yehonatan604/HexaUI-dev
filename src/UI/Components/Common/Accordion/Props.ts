import { TComponent } from "../../../../Data/Types/ComponentTypes/TComponent";

export type AccordionProps = TComponent<"div"> & {
    title: string;
    variant?: string;
};