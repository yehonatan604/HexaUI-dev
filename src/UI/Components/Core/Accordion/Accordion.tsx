import { useState } from "react";
import { IoIosArrowDown, IoIosArrowUp } from "react-icons/io";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import Flex from "../../Layout/Flex/Flex";
import { TAccordion } from "./types/TAccordion";
import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";

const Accordion = (props: TAccordion) => {
  const { options, children, ...componentProps } = props;
  const [isContentOpen, setIsContentOpen] = useState(false);
  const { colors } = useTheme();

  const bgColor = options.bgVariant || "transparent";
  const textColor = options?.textVariant || "gray-800";
  const borderColor = options.border?.variant || "gray-300";
  const borderRadius = options.border?.radius || "md";
  const border = options.border
    ? `border-2 border-${borderColor} rounded-${borderRadius}`
    : "";
  const titleColor = options.titleVariant || colors.secondaryLight;

  return (
    <Flex
      options={{
        direction: FlexDir.Col,
      }}
      className={`w-full ${border} p-4 gap-4`}
    >
      <Flex
        options={{
          justify: FlexTypes.SpaceBetween,
          align: FlexTypes.Center,
        }}
        className={`gap-4 ${bgColor} cursor-pointer w-full`}
        {...componentProps}
        onClick={() => setIsContentOpen(!isContentOpen)}
      >
        <h1 className={`text-xl text-${titleColor}`}>{options.title}</h1>
        <div className={`text-${textColor}`}>
          {isContentOpen ? <IoIosArrowUp /> : <IoIosArrowDown />}
        </div>
      </Flex>
      {isContentOpen && <div className={`text-${textColor}`}>{children}</div>}
    </Flex>
  );
};

export default Accordion;
