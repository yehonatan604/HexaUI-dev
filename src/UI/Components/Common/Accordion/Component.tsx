import { useState } from "react";
import { IoIosArrowDown, IoIosArrowUp } from "react-icons/io";
import { AccordionProps } from "./Props";
import useTheme from "../../../../Core/Hooks/useTheme";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { Variants } from "../../../../Data/Constants/Variants";
import Flex from "../../Layout/Flex/Component";

const Accordion = (props: AccordionProps) => {
  const { title, variant, children, ...componentProps } = props;
  const [isContentOpen, setIsContentOpen] = useState(false);
  const mode = useTheme().mode;

  const textColor = mode === "light" ? "text-gray-800" : "text-slate-200";

  const borderColor =
    variant === Variants.Standard
      ? "border-gray-300"
      : variant === Variants.Info
      ? "border-blue-300"
      : variant === Variants.Success
      ? "border-green-300"
      : variant === Variants.Warning
      ? "border-yellow-300"
      : variant === Variants.Failure
      ? "border-red-300"
      : variant === Variants.Lime
      ? "border-lime-300"
      : variant === Variants.Cyan
      ? "border-cyan-300"
      : variant === Variants.Purple
      ? "border-purple-300"
      : variant === Variants.Pink
      ? "border-pink-300"
      : "border-gray-300";
  return (
    <Flex
      direction={FlexDir.Col}
      className={`w-full border ${borderColor} rounded-md p-4 gap-4`}
    >
      <Flex
        justify={FlexTypes.SpaceBetween}
        align={FlexTypes.Center}
        className="gap-4 cursor-pointer w-full"
        {...componentProps}
        onClick={() => setIsContentOpen(!isContentOpen)}
      >
        <h1 className={`text-xl ${textColor}`}>{title}</h1>
        <div className={`${textColor}`}>
          {isContentOpen ? <IoIosArrowUp /> : <IoIosArrowDown />}
        </div>
      </Flex>
      {isContentOpen && <div className={`${textColor}`}>{children}</div>}
    </Flex>
  );
};

export default Accordion;
