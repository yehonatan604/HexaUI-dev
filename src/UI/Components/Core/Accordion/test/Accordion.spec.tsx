import { render, screen, fireEvent } from "@testing-library/react";
import { expect, test } from "vitest";
import Accordion from "../Accordion";

test("renders Accordion header correctly", () => {
  render(
    <Accordion options={{ title: "Accordion Header" }}>Accordion Content</Accordion>
  );
  expect(screen.getByText("Accordion Header")).toBeInTheDocument();
});

test("renders Accordion content when clicked", () => {
  render(
    <Accordion options={{ title: "Accordion Header" }}>Accordion Content</Accordion>
  );
  const header = screen.getByText("Accordion Header");
  fireEvent.click(header);
  expect(screen.getByText("Accordion Content")).toBeInTheDocument();
});

test("toggles content visibility on multiple clicks", () => {
  render(
    <Accordion options={{ title: "Accordion Header" }}>Accordion Content</Accordion>
  );
  const header = screen.getByText("Accordion Header");

  // Click to open
  fireEvent.click(header);
  expect(screen.getByText("Accordion Content")).toBeVisible();

  // Click again to close
  fireEvent.click(header);
  expect(screen.queryByText("Accordion Content")).not.toBeInTheDocument();
});

test("applies custom border color variant", () => {
  const { container } = render(
    <Accordion options={{ title: "Accordion Header", border: { variant: "blue-500" } }}>
      Accordion Content
    </Accordion>
  );
  expect(container.firstChild).toHaveClass("border-blue-500");
});

test("renders multiple accordions independently", () => {
  render(
    <>
      <Accordion options={{ title: "Accordion 1" }}>Content 1</Accordion>
      <Accordion options={{ title: "Accordion 2" }}>Content 2</Accordion>
    </>
  );

  // Open Accordion 1
  fireEvent.click(screen.getByText("Accordion 1"));
  expect(screen.getByText("Content 1")).toBeVisible();

  // Open Accordion 2
  fireEvent.click(screen.getByText("Accordion 2"));
  expect(screen.getByText("Content 2")).toBeVisible();
});

test("applies custom text color variant", () => {
  const { container } = render(
    <Accordion options={{ title: "Accordion Header", textVariant: "success" }}>
      Accordion Content
    </Accordion>
  );
  expect(container.firstChild.firstChild.lastChild).toHaveClass("text-success");
});

test("applies custom title color variant", () => {
  const { container } = render(
    <Accordion options={{ title: "Accordion Header", titleVariant: "approve" }}>
      Accordion Content
    </Accordion>
  );
  expect(container.firstChild.firstChild.firstChild).toHaveClass("text-approve");
});

test("applies custom background color variant", () => {
  const { container } = render(
    <Accordion options={{ title: "Accordion Header", bgVariant: "bg-green-500" }}>
      Accordion Content
    </Accordion>
  );
  expect(container.firstChild.firstChild).toHaveClass("bg-green-500");
});

test("applies custom border", () => {
  const { container } = render(
    <Accordion options={{ title: "Accordion Header", border: { radius: "xl" } }}>
      Accordion Content
    </Accordion>
  );
  expect(container.firstChild).toHaveClass("rounded-xl");
});
