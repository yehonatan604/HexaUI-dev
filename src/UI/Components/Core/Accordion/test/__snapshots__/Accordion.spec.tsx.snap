// Vitest Snapshot v1

exports[`Accordion Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col bg- text- justify-start items-start flex-nowrap w-full  p-4 gap-4"
  >
    <div
      class="flex flex-row bg- text- justify-between items-center flex-nowrap gap-4 transparent cursor-pointer w-full"
    >
      <h1
        class="text-xl text-secondary-l"
      >
        Accordion Header
      </h1>
      <div
        class="text-gray-800"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col bg- text- justify-start items-start flex-nowrap w-full  p-4 gap-4"
  >
    <div
      class="flex flex-row bg- text- justify-between items-center flex-nowrap gap-4 transparent cursor-pointer w-full"
    >
      <h1
        class="text-xl text-secondary-l"
      >
        Accordion Header
      </h1>
      <div
        class="text-gray-800"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
